### PR TITLE
Handle dupnoneed in forward mode

### LINKED ISSUE #392: Forward Mode DupNoNeed Bug
See: https://enzyme.mit.edu/explorer/#g:!((g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:65,endLineNumber:18,positionColumn:65,positionLineNumber:18,selectionStartColumn:65,selectionStartLineNumber:18,startColumn:65,startLineNumber:18),source:'%23include+%3Calgorithm%3E%0A%23include+%3Ccmath%3E%0A%23include+%3Cstring.h%3E%0A%23include+%3Cstdio.h%3E%0A%23include+%3Cassert.h%3E%0A%0Adouble+__enzyme_autodiff(void*,+...)%3B%0A%0Adouble+__enzyme_fwddiff(void*,+...)%3B%0A%0Adouble+f(double+x)%7B%0A++++return+x%3B%0A%7D%0A%0Aint+enzyme_dupnoneed%3B%0A%0Aint+main()+%7B%0A++++printf(%22%25f%5Cn%22,+__enzyme_fwddiff((void*)f,+enzyme_dupnoneed,+2.0,+1.0))%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:49.71110610861948,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang12-enzyme-main,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:eigen,ver:'340')),options:'-g0+-S+-emit-llvm+-O0+-ffast-math',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'clang+12+(f7c9fe5)+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:50.28889389138052,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',m:75.63913697326828,n:'0',o:'',t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+clang+12+(f7c9fe5)+(Compiler+%231)',t:'0')),header:(),l:'4',m:24.360863026731728,n:'0',o:'',s:0,t:'0')),l:'3',n:'0',o:'',t:'0')),version:4

```c
#include <algorithm>
#include <cmath>
#include <string.h>
#include <stdio.h>
#include <assert.h>

double __enzyme_autodiff(void*, ...);

double __enzyme_fwddiff(void*, ...);

double f(double x){
    return x;
}

int enzyme_dupnoneed;

int main() {
    printf("%f\n", __enzyme_fwddiff((void*)f, enzyme_dupnoneed, 2.0, 1.0));
}
```

### PR DESCRIPTION
**Pull request by tgymnich**
this fixes #392, we still need to look deeper into properly supporting dupnoneed.

### CODE REVIEW

**Comment by wsmoses** [comment]
Can you add a test?

**Comment by tgymnich** [comment]
included a test case covering this and #393 in the next PR.

**Comment by wsmoses** [comment]
Can you also include a calloc test for the multiple arg allocation function.

**Comment by tgymnich** [comment]
enzyme/test/Enzyme/ForwardMode/calloc.ll

**Comment by wsmoses** [comment]
While the file is named calloc, it doesn't actually check an allocation, no?

**Comment by tgymnich** [comment]
It was late late yesterday night 😅 I must've pasted from the wrong compiler explorer.

**Review by wsmoses — APPROVED** [review]
Please add the actual calloc test (and make sure there's still a dupnoneed test), otherwise LGTM.

File: enzyme/test/Enzyme/ForwardMode/calloc.ll [hunk 1]
```diff
@@ -0,0 +1,49 @@
+; RUN: %opt < %s %loadEnzyme -enzyme -enzyme-preopt=false -mem2reg -instsimplify -adce -simplifycfg -S | FileCheck %s
+
+
+@enzyme_dupnoneed = dso_local global i32 0, align 4
+
+define dso_local double @f(double %x) {
+entry:
+  %x.addr = alloca double, align 8
+  %data = alloca double*, align 8
+  store double %x, double* %x.addr, align 8
```
Comment (wsmoses) [inline enzyme/test/Enzyme/ForwardMode/calloc.ll:10]: For simplification, can you remove the extraneous alloas?

Also do you mind making one of the arguments of calloc passed into the function as an argument (to ensure the fix you added here of `getNewFromOriginal` doesn't accidentally pass if removed).

Comment (tgymnich) [reply enzyme/test/Enzyme/ForwardMode/calloc.ll]: done

**Pull request merged**